test/forkMainnet/LimitOrderSwap/Fill.t.sol

```diff
@@ -0,0 +1,527 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import { IUniswapRouterV2 } from "contracts/interfaces/IUniswapRouterV2.sol";
+import { ILimitOrderSwap } from "contracts/interfaces/ILimitOrderSwap.sol";
+import { Constant } from "contracts/libraries/Constant.sol";
+import { LimitOrder, getLimitOrderHash } from "contracts/libraries/LimitOrder.sol";
+import { BalanceSnapshot, Snapshot } from "test/utils/BalanceSnapshot.sol";
+import { LimitOrderSwapTest } from "test/forkMainnet/LimitOrderSwap/Setup.t.sol";
+
+contract FillTest is LimitOrderSwapTest {
+    using BalanceSnapshot for Snapshot;
+
+    address[] defaultAMMPath = [DAI_ADDRESS, USDT_ADDRESS];
+
+    function testFullyFillLimitOrder() public {
+        Snapshot memory takerTakerToken = BalanceSnapshot.take({ owner: taker, token: defaultOrder.takerToken });
+        Snapshot memory takerMakerToken = BalanceSnapshot.take({ owner: taker, token: defaultOrder.makerToken });
+        Snapshot memory makerTakerToken = BalanceSnapshot.take({ owner: defaultOrder.maker, token: defaultOrder.takerToken });
+        Snapshot memory makerMakerToken = BalanceSnapshot.take({ owner: defaultOrder.maker, token: defaultOrder.makerToken });
+        Snapshot memory recTakerToken = BalanceSnapshot.take({ owner: recipient, token: defaultOrder.takerToken });
+        Snapshot memory recMakerToken = BalanceSnapshot.take({ owner: recipient, token: defaultOrder.makerToken });
+        Snapshot memory feeCollectorBal = BalanceSnapshot.take({ owner: feeCollector, token: defaultOrder.makerToken });
+
+        uint256 fee = (defaultOrder.makerTokenAmount * defaultFeeFactor) / Constant.BPS_MAX;
+
+        vm.expectEmit(true, true, true, true);
+        emit LimitOrderFilled(
+            getLimitOrderHash(defaultOrder),
+            taker,
+            defaultOrder.maker,
+            defaultOrder.takerToken,
+            defaultOrder.takerTokenAmount,
+            defaultOrder.makerToken,
+            defaultOrder.makerTokenAmount - fee,
+            fee,
+            recipient
+        );
+
+        vm.prank(taker);
+        limitOrderSwap.fillLimitOrder({ order: defaultOrder, makerSignature: defaultMakerSig, takerParams: defaultTakerParams });
+
+        takerTakerToken.assertChange(-int256(defaultOrder.takerTokenAmount));
+        takerMakerToken.assertChange(int256(0));
+        makerTakerToken.assertChange(int256(defaultOrder.takerTokenAmount));
+        makerMakerToken.assertChange(-int256(defaultOrder.makerTokenAmount));
+        recTakerToken.assertChange(int256(0));
+        recMakerToken.assertChange(int256(defaultOrder.makerTokenAmount - fee));
+        feeCollectorBal.assertChange(int256(fee));
+    }
+
+    function testFullyFillLimitOrderUsingAMM() public {
```

Comment:
Add a test for strategy failing to deliver enough taker token amount?

Reply:
Actually it depends on the logic of strategy contract. A simple case could be that taker doesn't have any taker tokens and only relies on strategy contract. Then if strategy contract not returning enough to taker then the collecting taker token from taker will fail (the last `transferFrom`).

Added in 2b9cb7d.